// clicked/clicked.js
/**
 * Javascript: create click event for both mouse and touch
 * @example
 *
 * import clicked from 'clicked';
 * // or var clicked = require('clicked');
 *
 *  function handleClick()
 *  {
 *      console.log('I was clicked.');
 *  }
 *
 *  var div = document.getElementById('clickme');
 *  clicked(div, handleClick, {thresshold: 15});
 *
 */

/**
 * @param {HTMLElement} element
 * @param {function} callback called after click: callback(event, options.args)
 * @param {object} [options]
 * @param {number} [options.thresshold=10] if touch moves threshhold-pixels then the touch-click is cancelled
 * @param {*} [options.args] arguments for callback function
 * @returns {object} object
 * @returns {function} object.disable() - disables clicked
 * @returns {function} object.enable() - enables clicked after disable() is called
 */
function clicked(element, callback, options)
{
    function touchstart(e)
    {
        if (disabled)
        {
            return;
        }
        if (e.touches.length === 1)
        {
            lastX = e.changedTouches[0].screenX;
            lastY = e.changedTouches[0].screenY;
            down = true;
        }
    }

    function pastThreshhold(x, y)
    {
        return Math.abs(lastX - x) > threshhold || Math.abs(lastY - y) > threshhold;
    }

    function touchmove(e)
    {
        if (!down || e.touches.length !== 1)
        {
            touchcancel();
            return;
        }
        var x = e.changedTouches[0].screenX;
        var y = e.changedTouches[0].screenY;
        if (pastThreshhold(x, y))
        {
            touchcancel();
        }
    }

    function touchcancel()
    {
        down = false;
    }

    function touchend(e)
    {
        if (down)
        {
            e.preventDefault();
            callback(e, options.args);
        }
    }

    function mouseclick(e)
    {
        if (!disabled)
        {
            callback(e, options.args);
        }
    }

    options = options || {};
    var down, lastX, lastY, disabled;
    var threshhold = options.thresshold || 10;

    element.addEventListener('click', mouseclick);
    element.addEventListener('touchstart', touchstart, { passive: true });
    element.addEventListener('touchmove', touchmove, { passive: true });
    element.addEventListener('touchcancel', touchcancel);
    element.addEventListener('touchend', touchend);

    return {
        disable: function () { disabled = true; },
        enable: function () { disabled = false; }
    };
}

module.exports = clicked;
